### ISSUE TITLE
Feature Request: Path translation in WSL and Git Bash

### ISSUE THREAD

**Issue by Forinil** [microsoft/terminal#1772]
# Summary of the new feature/enhancement

It would be useful if the path pasted from clipboard to a shell was translated to a format that said shell understands if it differs form native windows format.

For example, if user pastes `C:\Windows\System\` to PowerShell or cmd it remains unchanged, but in WSL it changes to `/mnt/c/Windows/System` and in Git Bash to `/c/Windows/System`.

**Comment by eberkund** [comment]
I wonder if this would be implementable in a generic way such as by having two settings like this for Git bash:
```
"pathConversionForwardSlashes": true,
"pathConversionRemovePrefix": "C:\",
"pathConversionReplacePrefix": "/mnt/c/"
```

**Comment by Forinil** [comment]
The prefixes would have to be regular expressions to acount for the fact that C is not the only possible drive letter.

Maybe something like:
```regexp
"pathConversionExpression": "|([A-Za-z])\(.*)|/mnt/\L$1/$2|"
```

where $1, $2, ... are capture groups and \L is a lowercase operator (like in Notepad++).

**Comment by eberkund** [comment]
Yeah you're right, I guess full blown regex would be a better solution at that point.

**Comment by ToadKing** [comment]
A better solution IMO would be to use the path conversion commands command built into WSL and Cygwin/MSYS2/Git Bash/etc. to convert it for us. So for WSL you would have:
```json
"pathConversionCmd": "wsl wslpath -a $1"
```
And for Git Bash you could have :
```json
"pathConversionCmd": "\"C:\\Program Files\\Git\\usr\\bin\\cygpath.exe\" -a $1"
```

**Comment by eleniums** [comment]
Any traction on this issue? It seems small, but this is actually a major roadblock to me switching from cmder to Windows Terminal.

**Comment by UnsnappedKhan8** [comment]
Sorry about that

**Comment by Daniel599** [comment]
> A better solution IMO would be to use the path conversion commands command built into WSL and Cygwin/MSYS2/Git Bash/etc. to convert it for us. So for WSL you would have:
> 
> ```json
> "pathConversionCmd": "wsl wslpath -a $1"
> ```
> 
> And for Git Bash you could have :
> 
> ```json
> "pathConversionCmd": "\"C:\\Program Files\\Git\\usr\\bin\\cygpath.exe\" -a $1"
> ```

I was looking into implementing this, I really like the idea but also I started thinking;
what about security issues? that config field will allow anyone to run any executable by hijacking the settings file (even with admin permissions if the terminal was opened that way).
Should I do the regex instead? or detect it's `wsl` and support only that?
@zadjii-msft @DHowett-MSFT what do you guys think?

**Comment by zadjii-msft** [comment]
You know what, we've got a team sync later today, I'll bring this up. We had a plan for WSL distros that would just work like magic, but that would _not_ work at all for mingw profiles. Let's see if we can come up with something.

I don't _love_ the idea of putting an arbitrary exe into the settings file itself. It's definitely a possible security hole. We might be able to hack around this temporarily with 
* A. making this an `experimental.` setting, in case we come up with a better plan
* B. letting this only work in unelevated windows. I know that's annoying, but prevents an escalation-of-privilege whereby an unelevated `malicious.exe` from changing your `settings.json` so that `pathConversionCmd` is set to `malicious.exe` instead

**Comment by Daniel599** [comment]
@zadjii-msft any update in that matter?
In the meanwhile I think I`ll try to implement #10073, as it seems to use working-dir to convert path so maybe those two issues are ortogonale

**Comment by zadjii-msft** [comment]
My apologies. A bunch of the team was out sick last week, so we ended up just not having the meeting 😬 Hopefully everyone's feeling better this week (and we're not to preoccupied with last-minute //Build fires)

**Comment by zadjii-msft** [comment]
notes from team sync: someone recalled vaguely that ConEmu has a solution for this that doesn't involve launching another exe to resolve that path. Maybe we should take a look at what they do, and see if that makes sense to emulate.

**Comment by skyline75489** [comment]
Here's the documentation of Shell working dir in ConEmu for reference: https://conemu.github.io/en/ShellWorkDir.html 

I see `ConEmuC.exe` which is a native win32 application that seems to handle path resolution.

Also tips for @Daniel599: Dustin no longer uses the account `DHowett-MSFT` for this project. He's now @DHowett .

**Comment by shinayser** [comment]
Any update on that issue?

**Comment by zadjii-msft** [comment]
Nope. We'll make sure to update this thread when there is. In the meantime, might I recommend the Subscribe button?
![image](https://user-images.githubusercontent.com/18356694/91237459-5cbb0c80-e700-11ea-9347-b9b1ec2813b1.png)
That way you'll be notified of any updates to this thread, without needlessly pinging everyone on this thread ☺️


* #11625 added support for translating drag/dropped WSL paths. 
* We'd still need to do something similar if the clipboard contained a path (this issue)
* #3158 is for tracking "the shell told us it was in {WSL path}, which we can actually reverse engineer to {Windows path}"
* There's still no good solution for cygwin/msys paths. That's not something we control, so we're not really planning on adding tighter integration with that. Throwing arbitrary exe's in the settings to have the Terminal execute seems generally like a bad idea.

**Comment by zadjii-msft** [comment]
We certainly don't have the capacity to get to this ourselves any time soon, sorry 😕. If someone wants to help sort out the bits of this to enable this for WSL distros, I'd be more than capable to give pointers of where I'd start. The "use an exe to translate the path" seems a bit more... risqué, but the "I know the thing running in this terminal is _this_ WSL so we can translate the path {like so}" seems pretty doable.

**Comment by MattBussing** [comment]
Thank for for the quick reply! I don't plan on working on it at the moment.

**Comment by sebolio** [comment]
i'm pretty sure to have seen this at some point some time ago, i dragged a folder to terminal with ubuntu open on it, and it replaced the path to the corresponding `/mnt/c` path, but i can't see it working now 🤔

**Comment by zadjii-msft** [comment]
@sebolio This only works for panes created for profiles with a `Microsoft.Terminal.WSL` `source` set. There's a set of `"source": "CanonicalGroupLimited.Ubuntu*` profiles floating around, where that won't work. That's vaguely tracked in #14233, #14003 and #12961

**Comment by garretzou** [comment]
We're in 2024 now, but Windows remains unable to automatically transform its own paths into Linux-style paths when pasting.

**Comment by ParadoxV5** [comment]
#11625 detects whether the profile is WSL buy checking for `"source": "Windows.Terminal.Wsl"` in the configs (which isn’t exposed to the Settings UI because why should we?). This means manually-created WSL profiles, such as my Interactive Ruby (`wsl irb`), don’t get the same VIP treatment.

Please, either
* Detect WSL from the start-up command (`/\Awsl(.exe)?/`) rather than that arcane setting
* Make path translations manually customizable – users get to decide whether to translate or not.
  * The automatic profile from WSL installation will instead be created with this config set to translating Windows paths to Linux.

**Comment by DHowett** [comment]
> which isn’t exposed to the Settings UI because why should we?

Because that setting literally tracks where the profile came from.

In general, detecting _what a profile is running_ versus _where a profile came from_ is difficult and annoying.

Even these proposed solutions don't work for the case where you run `wsl` from within PowerShell, or run `ubuntu2204.exe` which is a wrapper around the WSL API.

Yes, path translations should be manually customizable. That's why this issue is still open, after all! I'd rather have _some_ progress than _no_ progress.

**Comment by ParadoxV5** [comment]
> Even these proposed solutions don't work for the case where you run `wsl` from within PowerShell, or run `ubuntu2204.exe` which is a wrapper around the WSL API.

Is it possible for the terminal to know what (sub)process it is currently hosting?

**Comment by DHowett** [comment]
> Is it possible for the terminal to know what (sub)process it is currently hosting?

To an extent, yes :)

We have it on the books to track what we're hosting so that we can display better error messages and do profile matching. However, it'll break down for "opaque" remoting protocols like SSH and all of WSL (since processes on the Linux side don't present as processes on the Windows side) ☹️ 

Still, that's better than nothing!

**Comment by MichalPetryka** [comment]
Wouldn't it be enough to just introduce a separate `"unixPaths": "true"` setting for it for most usecases?

**Comment by ParadoxV5** [comment]
> However, it'll break down for "opaque" remoting protocols like SSH and all of WSL (since processes on the Linux side don't present as processes on the Windows side) ☹️
> 
> Still, that's better than nothing!

At least enough for the terminal to know that it’s running something WSL!
(Probably not enough to know that it’s hosting a Linux through SSH, though.)

**Comment by driver1998** [comment]
We already use VT to let shells announce their CWD, why not use VT to let shells announce what they actually are?

**Comment by weberjn** [comment]
For MSYS2 it also works putting single quotes around the dropped windows path: 'C:\Windows', MSYS2 correctly translates that.
So, Windows Terminal would need an option [x] put quotes around dropped paths.

**Comment by DHowett** [comment]
Oh hey, I didn't realize this bug existed!

We added a new "path translation style" option that mostly addresses this.

![Image](https://github.com/user-attachments/assets/310512d5-dc0f-436f-bdc6-8f1bc4e1d955)

It doesn't afford customization, but it does cover a few of the common use cases.

We've also worked with the Git for Windows folks to make sure it's set up for them by default.

**Comment by mrhappytech** [comment]
I just tried preview version Microsoft.WindowsTerminalPreview_1.23.10353.0_x64.zip and it doesn't work correctly with Bash shell. 
I configured it at Bash -> Advanced -> Path translation like MSYS2 (C:\ -> /c)

When using right mouse button or ctrl+v the result is:

```
$ cd c:\Program Files (x86)\
```

instead of correct

```
$ cd '/c/Program Files (x86)/'
or
$ cd /c/MyProgramFiles/
```

If I want to add workaround inside .bashrc as cdwin function:

```
$ cdwin c:\MyProgramFiles\
>
I receive the variable --> c:MyProgramFiles  // it doesn't contain \ characters
/c:MyProgramFiles
bash: cd: /c:MyProgramFiles: No such file or directory
```

Note:

If I use 

```
$ cd c:\MyProgramFiles\  //press enter
> //press enter
```

then it will switch folder correctly

`me@pc mini /c/MyProgramFiles `

but for 

```
$ cd c:\Program Files (x86)\
bash: syntax error near unexpected token `('
```

I am surprised that such basic functionality is still not there.

**Comment by DHowett** [comment]
Sorry. Paths **stored as text** on the clipboard are still just considered text. We have not changed how **text** is translated.

Paths generated via drag/drop and copying full file objects (from file explorer) to the clipboard are translated.

**Comment by vaclavsramek** [comment]
I got it. Thanks for explanation.

I am using this as a workaround:

`# cd to bash path from windows path
function cdw() {
    echo "I receive the variable --> $1"
    path=$(sed -e 's#:##g' -e 's#\\#/#g' <<< "/$1")
	  echo "Switching to folder --> $path"
    cd "$path"
}

# windows path to bash
function w2b() {
    path=$(sed -e 's#:##g' -e 's#\\#/#g' <<< "/$1")
	echo "$path"
}`
